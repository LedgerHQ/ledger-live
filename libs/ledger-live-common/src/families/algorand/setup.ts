// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
