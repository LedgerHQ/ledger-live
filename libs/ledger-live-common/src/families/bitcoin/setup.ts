// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { firstValueFrom, from } from "rxjs";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import Btc from "@ledgerhq/hw-app-btc";
import { createBridges } from "@ledgerhq/coin-bitcoin/bridge/js";
import type { SignerContext } from "@ledgerhq/coin-bitcoin/signer";
import makeCliTools from "@ledgerhq/coin-bitcoin/cli-transaction";
import bitcoinResolver from "@ledgerhq/coin-bitcoin/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-bitcoin/hw-signMessage";
import { Transaction } from "@ledgerhq/coin-bitcoin/types";
import { GetAddressOptions, Resolver } from "../../hw/getAddress/types";
import { withDevice } from "../../hw/deviceAccess";
import { startSpan } from "../../performance";
import { GetAddressFn } from "@ledgerhq/coin-framework/bridge/getAddressWrapper";
import { getCurrencyConfiguration } from "../../config";
import { BitcoinConfigInfo } from "@ledgerhq/coin-bitcoin/lib/config";
import { SignMessage } from "../../hw/signMessage/types";

const createSigner = (transport: Transport, currency: CryptoCurrency) => {
  return new Btc({ transport, currency: currency.id });
};

const signerContext: SignerContext = <T>(
  deviceId: string,
  crypto: CryptoCurrency,
  fn: (signer: Btc) => Promise<T>,
): Promise<T> =>
  firstValueFrom(
    withDevice(deviceId)((transport: Transport) => from(fn(createSigner(transport, crypto)))),
  );

const getCurrencyConfig = (currency: CryptoCurrency) => {
  return { info: getCurrencyConfiguration<BitcoinConfigInfo>(currency) };
};

const perfLogger = {
  startSpan,
};

const bridge: Bridge<Transaction> = createBridges(signerContext, perfLogger, getCurrencyConfig);

export function createMessageSigner(): SignMessage {
  return (transport, account, messageData) => {
    const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
    return signMessage(signerContext)("", account, messageData);
  };
}

const messageSigner = {
  signMessage: createMessageSigner(),
};

const resolver: Resolver = (
  transport: Transport,
  addressOpt: GetAddressOptions,
): ReturnType<GetAddressFn> => {
  const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
  return bitcoinResolver(signerContext)("", addressOpt);
};

const cliTools = makeCliTools();

export { bridge, cliTools, resolver, messageSigner, signerContext };
