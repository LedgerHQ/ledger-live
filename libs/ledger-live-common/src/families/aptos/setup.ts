// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
