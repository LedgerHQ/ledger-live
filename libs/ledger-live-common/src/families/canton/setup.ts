// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-canton/bridge/index";
import { getEnv } from "@ledgerhq/live-env";
import Transport from "@ledgerhq/hw-transport";
import { CantonSigner } from "@ledgerhq/coin-canton";
import cantonResolver from "@ledgerhq/coin-canton/signer";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-canton/test/cli";
import { CantonCoinConfig } from "@ledgerhq/coin-canton/config";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-canton/types";
import { LegacySignerCanton } from "@ledgerhq/live-signer-canton";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import cantonBridgeMock from "./bridge/mock";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "../../currencies";

const createSigner: CreateSigner<CantonSigner> = (transport: Transport) => {
  return new LegacySignerCanton(transport);
};

const getCurrencyConfig = () => {
  // Use devnet for development/testing
  const currencyId = getEnv("MOCK") ? "canton_network_devnet" : "canton_network_devnet";
  return getCurrencyConfiguration<CantonCoinConfig>(getCryptoCurrencyById(currencyId));
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = getEnv("MOCK")
  ? cantonBridgeMock
  : createBridges(executeWithSigner(createSigner), getCurrencyConfig);

const resolver: Resolver = createResolver(createSigner, cantonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
