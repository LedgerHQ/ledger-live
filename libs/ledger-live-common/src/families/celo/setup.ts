// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-celo/bridge";
import makeCliTools from "@ledgerhq/coin-celo/cli-transaction";
import CeloResolver from "@ledgerhq/coin-celo/hw-getAddress";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import Celo from "@ledgerhq/hw-app-celo";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import { Transaction, CeloAccount } from "./types";
import { CeloSigner } from "@ledgerhq/coin-celo/signer";

const createSigner: CreateSigner<CeloSigner> = (transport: Transport) => new Celo(transport);

const bridge: Bridge<Transaction, CeloAccount> = createBridges(executeWithSigner(createSigner));

const resolver: Resolver = createResolver(createSigner, CeloResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
