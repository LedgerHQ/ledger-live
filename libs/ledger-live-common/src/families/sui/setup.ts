// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { SuiAccount, TransactionStatus, createBridges, type Transaction } from "@ledgerhq/coin-sui";
import Transport from "@ledgerhq/hw-transport";
import Sui from "@ledgerhq/hw-app-sui/lib/Sui";
import type { Bridge } from "@ledgerhq/types-live";
import { SuiCoinConfig } from "@ledgerhq/coin-sui/config";
import { type SuiSigner } from "@ledgerhq/coin-sui/types/signer";
import suiResolver from "@ledgerhq/coin-sui/signer/index";
// import makeCliTools, { type CliTools } from "@ledgerhq/coin-sui/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";

const createSigner: CreateSigner<SuiSigner> = (transport: Transport) => {
  return new Sui(transport);
};

const sui = getCryptoCurrencyById("sui");
const getCurrencyConfig = (): SuiCoinConfig => {
  return getCurrencyConfiguration(sui);
};

const bridge: Bridge<Transaction, SuiAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCurrencyConfig,
);

const resolver: Resolver = createResolver(createSigner, suiResolver);

// const cliTools: CliTools = makeCliTools();

export {
  bridge,
  // cliTools,
  resolver,
};
