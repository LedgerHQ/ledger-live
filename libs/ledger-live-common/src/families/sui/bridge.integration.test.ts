import "../../__tests__/test-helpers/setup";

import { dataset } from "@ledgerhq/coin-sui/test/index";
import { testBridge } from "../../__tests__/test-helpers/bridge";

testBridge(dataset);
