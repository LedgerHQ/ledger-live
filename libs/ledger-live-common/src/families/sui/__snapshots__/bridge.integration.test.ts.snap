// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sui currency bridge scanAccounts sui seed 1 1`] = `
[
  {
    "balance": "102696240",
    "currencyId": "sui",
    "derivationMode": "sui",
    "freshAddress": "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
    "freshAddressPath": "44'/784'/0'/0'/0'",
    "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
    "index": 0,
    "operationsCount": 21,
    "pendingOperations": [],
    "seedIdentifier": "e5fb8c817b97e9e9aec677b9e822e8a60c7471624f36b701c94560b88195a8e7",
    "spendableBalance": "102696240",
    "suiResources": {},
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
  },
  {
    "balance": "0",
    "currencyId": "sui",
    "derivationMode": "sui",
    "freshAddress": "0xba7080172a6d957b9ed2e3eb643529860be963cf4af896fb84f1cde00f46b561",
    "freshAddressPath": "44'/784'/1'/0'/0'",
    "id": "js:2:sui:0xba7080172a6d957b9ed2e3eb643529860be963cf4af896fb84f1cde00f46b561:sui",
    "index": 1,
    "operationsCount": 0,
    "pendingOperations": [],
    "seedIdentifier": "e5fb8c817b97e9e9aec677b9e822e8a60c7471624f36b701c94560b88195a8e7",
    "spendableBalance": "0",
    "suiResources": {},
    "swapHistory": [],
    "syncHash": undefined,
    "used": false,
  },
]
`;

exports[`sui currency bridge scanAccounts sui seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "2A4xTFFpdUYtW92GBSSLALJYymGjEViAiKcyUdUV7u9F",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "2A4xTFFpdUYtW92GBSSLALJYymGjEViAiKcyUdUV7u9F",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-2A4xTFFpdUYtW92GBSSLALJYymGjEViAiKcyUdUV7u9F-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "19745060210",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "2U4dyfZTn5FwQJPWt5DKhGsbVewFtbAA43obHdFgyd1W",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "2U4dyfZTn5FwQJPWt5DKhGsbVewFtbAA43obHdFgyd1W",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-2U4dyfZTn5FwQJPWt5DKhGsbVewFtbAA43obHdFgyd1W-OUT",
      "recipients": [
        "0x294f9f370f4ed53d53702b8de1b4f0d73980806d7148ad64edc0be80192f5107",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "18402932460",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "3ZE1Kfo9zohdyRLA5fzcoULeYEhnYJXiEpudGrvGe3w5",
      "blockHeight": 5,
      "extra": {},
      "fee": "-217360",
      "hasFailed": false,
      "hash": "3ZE1Kfo9zohdyRLA5fzcoULeYEhnYJXiEpudGrvGe3w5",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-3ZE1Kfo9zohdyRLA5fzcoULeYEhnYJXiEpudGrvGe3w5-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x6e143fe0a8ca010a86580dafac44298e5b1b7d73efc345356a59a15f0d7824f0",
      ],
      "type": "IN",
      "value": "21600000000",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "4ykxTJUrTEJswRZWafgN4m3GgeXavfPNJuBzn8p93bHo",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "4ykxTJUrTEJswRZWafgN4m3GgeXavfPNJuBzn8p93bHo",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-4ykxTJUrTEJswRZWafgN4m3GgeXavfPNJuBzn8p93bHo-OUT",
      "recipients": [
        "0x9b3a01bb3b8c86bdeab6afc28b19390fab898131d40696bcee643db14d51b156",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "20195822240",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "6p9XjoTwZ81o9xTkodgswDbETbrN5Lkv1b2shBN4p99s",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "6p9XjoTwZ81o9xTkodgswDbETbrN5Lkv1b2shBN4p99s",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-6p9XjoTwZ81o9xTkodgswDbETbrN5Lkv1b2shBN4p99s-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x9b3a01bb3b8c86bdeab6afc28b19390fab898131d40696bcee643db14d51b156",
      ],
      "type": "IN",
      "value": "100000000",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "7bsW7Aj5S6iQi9RxSdjaCPkwZcBbuUAfGAXFb5Ydn5Ag",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "7bsW7Aj5S6iQi9RxSdjaCPkwZcBbuUAfGAXFb5Ydn5Ag",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-7bsW7Aj5S6iQi9RxSdjaCPkwZcBbuUAfGAXFb5Ydn5Ag-OUT",
      "recipients": [
        "0x0bb481f4ad94c24ec7e9ac4c04a55c12fcc64fd0009c7e4db49471defba84c86",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "21553456960",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "7pQVzNm9eCK974n4bzLnVte2Rc8EmMUb9ehvfpsa3unC",
      "blockHeight": 5,
      "extra": {},
      "fee": "1737880",
      "hasFailed": false,
      "hash": "7pQVzNm9eCK974n4bzLnVte2Rc8EmMUb9ehvfpsa3unC",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-7pQVzNm9eCK974n4bzLnVte2Rc8EmMUb9ehvfpsa3unC-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "13918448260",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "7YAS8kkE3piqopdWeQwHmAp3MRgumS8KuqFxvxQpt8oX",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "7YAS8kkE3piqopdWeQwHmAp3MRgumS8KuqFxvxQpt8oX",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-7YAS8kkE3piqopdWeQwHmAp3MRgumS8KuqFxvxQpt8oX-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "18402932390",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "9FZcUokehQA93FZo8pjHyqCFzTFjaMF7ZNE6Sa1ZPXDW",
      "blockHeight": 5,
      "extra": {},
      "fee": "760760",
      "hasFailed": false,
      "hash": "9FZcUokehQA93FZo8pjHyqCFzTFjaMF7ZNE6Sa1ZPXDW",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-9FZcUokehQA93FZo8pjHyqCFzTFjaMF7ZNE6Sa1ZPXDW-OUT",
      "recipients": [
        "0x6e143fe0a8ca010a86580dafac44298e5b1b7d73efc345356a59a15f0d7824f0",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "21538185960",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "Bgdvga7F6755F6yiTN4mSHwTpHTTDrBN5rLzChzvfbUB",
      "blockHeight": 5,
      "extra": {},
      "fee": "-7065200",
      "hasFailed": false,
      "hash": "Bgdvga7F6755F6yiTN4mSHwTpHTTDrBN5rLzChzvfbUB",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-Bgdvga7F6755F6yiTN4mSHwTpHTTDrBN5rLzChzvfbUB-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "19566792820",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "Bs7jiagEu69RpZt7By8PgNasLZhMfPbuD79yRTqrszag",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "Bs7jiagEu69RpZt7By8PgNasLZhMfPbuD79yRTqrszag",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-Bs7jiagEu69RpZt7By8PgNasLZhMfPbuD79yRTqrszag-OUT",
      "recipients": [
        "0xa1c80990a70a9674556a97838fa3d5814cd1f064c778be019804b19d9f01c8ec",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "24275847660",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "Cj6ghzCygwqpwpcxeG2TNKdtXkFAEd6VSVmnMcaqnA25",
      "blockHeight": 5,
      "extra": {},
      "fee": "1737880",
      "hasFailed": false,
      "hash": "Cj6ghzCygwqpwpcxeG2TNKdtXkFAEd6VSVmnMcaqnA25",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-Cj6ghzCygwqpwpcxeG2TNKdtXkFAEd6VSVmnMcaqnA25-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "16941487270",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "CqPV2f1f1fbNC1LHbbZxJTNYHPZrzn1KycB6rCFpMpjX",
      "blockHeight": 5,
      "extra": {},
      "fee": "1738880",
      "hasFailed": false,
      "hash": "CqPV2f1f1fbNC1LHbbZxJTNYHPZrzn1KycB6rCFpMpjX",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-CqPV2f1f1fbNC1LHbbZxJTNYHPZrzn1KycB6rCFpMpjX-OUT",
      "recipients": [
        "0x680dbfb4bd33c276b0192126fc79ddddfbb551e83ddf76405409209ceae30717",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "9001738880",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "DhXKA85ru8y2kAXpaEbzhMJD2nrzKNBw4Lfxq2EmGxqt",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "DhXKA85ru8y2kAXpaEbzhMJD2nrzKNBw4Lfxq2EmGxqt",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-DhXKA85ru8y2kAXpaEbzhMJD2nrzKNBw4Lfxq2EmGxqt-OUT",
      "recipients": [
        "0xdb3b3f123377fd6e4d93d5f1b3dfb9b004589ee5aeabfd54c382b11fedec2f57",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "9000759760",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "EF1ro2sgEzo8LnPhWncV4pcUxKEbfkpTLekBowDZ5PCm",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "EF1ro2sgEzo8LnPhWncV4pcUxKEbfkpTLekBowDZ5PCm",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-EF1ro2sgEzo8LnPhWncV4pcUxKEbfkpTLekBowDZ5PCm-OUT",
      "recipients": [
        "0xbc597145c42d720c5203425c3bb1195b5ade319d30f8a58a07c94090709ed5fb",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "21859175760",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "EhsMAo4u2GSg5y6TEAx1vnUPibzBpUWJb4yQcrFJC4Ux",
      "blockHeight": 5,
      "extra": {},
      "fee": "760760",
      "hasFailed": false,
      "hash": "EhsMAo4u2GSg5y6TEAx1vnUPibzBpUWJb4yQcrFJC4Ux",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-EhsMAo4u2GSg5y6TEAx1vnUPibzBpUWJb4yQcrFJC4Ux-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x9b3a01bb3b8c86bdeab6afc28b19390fab898131d40696bcee643db14d51b156",
      ],
      "type": "IN",
      "value": "15000000000",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "FQ4Wj6ToRG7GgvCSs98L9YTUQN6pzGZLi2y49SUTFbNs",
      "blockHeight": 5,
      "extra": {},
      "fee": "-218360",
      "hasFailed": false,
      "hash": "FQ4Wj6ToRG7GgvCSs98L9YTUQN6pzGZLi2y49SUTFbNs",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-FQ4Wj6ToRG7GgvCSs98L9YTUQN6pzGZLi2y49SUTFbNs-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "24275847710",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "FWftMWxSbk5ZfpVpAW3bPmP2wZjHG6Ei3XD6dBC16bJ1",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "FWftMWxSbk5ZfpVpAW3bPmP2wZjHG6Ei3XD6dBC16bJ1",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-FWftMWxSbk5ZfpVpAW3bPmP2wZjHG6Ei3XD6dBC16bJ1-OUT",
      "recipients": [
        "0x02600aefb4cc6df7051058f0c77d39f19927c16c083b293b8bc8383f1ff868ab",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "19745060160",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "FwkKBfz8vqKLrPNjqWUz26PamUehgMKkFkYaoEMxuCsz",
      "blockHeight": 5,
      "extra": {},
      "fee": "1738880",
      "hasFailed": false,
      "hash": "FwkKBfz8vqKLrPNjqWUz26PamUehgMKkFkYaoEMxuCsz",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-FwkKBfz8vqKLrPNjqWUz26PamUehgMKkFkYaoEMxuCsz-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "15542622080",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "GSDowHDQQ7tyQhwbqBhW6awKFxwjirhWkUBbWWyDdiWY",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "GSDowHDQQ7tyQhwbqBhW6awKFxwjirhWkUBbWWyDdiWY",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-GSDowHDQQ7tyQhwbqBhW6awKFxwjirhWkUBbWWyDdiWY-IN",
      "recipients": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "senders": [
        "0x861284f9839a0334f8bbdbc4d9b254769ca74a5139636de9785ab2b7814dca9c",
      ],
      "type": "IN",
      "value": "22583245100",
    },
    {
      "accountId": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui",
      "blockHash": "XWS3cmmihyiReNHY6k24ucti8YzKdMQAAzPJbnLx9x6",
      "blockHeight": 5,
      "extra": {},
      "fee": "759760",
      "hasFailed": false,
      "hash": "XWS3cmmihyiReNHY6k24ucti8YzKdMQAAzPJbnLx9x6",
      "id": "js:2:sui:0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435:sui-XWS3cmmihyiReNHY6k24ucti8YzKdMQAAzPJbnLx9x6-OUT",
      "recipients": [
        "0xb194c1eeab00f044e5c64df4423795f2cc8a6cf804cb9280590711096660c3d4",
      ],
      "senders": [
        "0x117527fdcf2f5f6e82fa499c0398def2643550d63f7e8311245d75f4635f0435",
      ],
      "type": "OUT",
      "value": "22000759760",
    },
  ],
  [],
]
`;
